src/v/s3/client.cc

```diff
@@ -320,6 +315,108 @@ request_creator::make_delete_object_request(
     return header;
 }
 
+struct delete_objects_body : public ss::data_source_impl {
+    ss::temporary_buffer<char> data;
+    explicit delete_objects_body(std::string_view body) noexcept
+      : data{body.data(), body.size()} {}
+    auto get() -> ss::future<ss::temporary_buffer<char>> override {
+        return ss::make_ready_future<ss::temporary_buffer<char>>(
+          std::exchange(data, {}));
+    }
+};
+
+result<std::tuple<http::client::request_header, ss::input_stream<char>>>
+request_creator::make_delete_objects_request(
+  const s3::bucket_name& name, std::span<const object_key> keys) {
+    // https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObjects.html
+    // will generate this request:
+    //
+    // POST /?delete HTTP/1.1
+    // Host: <Bucket>.s3.amazonaws.com
+    // Content-MD5: <Computer from body>
+    // Authorization: <applied by _requestor>
+    // Content-Length: <...>
+    //
+    // <?xml version="1.0" encoding="UTF-8"?>
+    // <Delete>
+    //     <Object>
+    //         <Key>object_key</Key>
+    //     </Object>
+    //     <Object>
+    //         <Key>object_key</Key>
+    //     </Object>
+    //      ...
+    //     <Quiet>true</Quiet>
+    // </Delete>
+    //
+    // note:
+    //  - Delete.Quiet true will generate a response that reports only failures
+    //  to delete or errors
+    //  - the actual xml might not be pretty-printed
+    //  - with clang15 and ranges, xml generation could be a one-liner + a
+    //  custom formatter for xml escaping
+
+    constexpr static auto s3_delete_objects_max_keys = 1000;
```

Comment:
Could you please move this to the header file? The code that will use the request should be able to use it to build a list of objects to delete.

Reply:
Good call